cli/src/commands/sign.rs

```diff
@@ -0,0 +1,144 @@
+    /// What revision(s) to sign
+    #[arg(
+        long, short,
+        value_name = "REVSETS",
+        add = ArgValueCandidates::new(complete::mutable_revisions),
+    )]
+    revisions: Vec<RevisionArg>,
+}
+
+pub fn cmd_sign(ui: &mut Ui, command: &CommandHelper, args: &SignArgs) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    if command.settings().signing_backend()?.is_none() {
+        return Err(user_error_with_hint(
+            "No signing backend configured.",
+            "For configuring a signing backend, see https://jj-vcs.github.io/jj/latest/config/#commit-signing",
+        ));
+    }
+
+    let is_authored_by_me =
+        |commit: &Commit| -> bool { commit.author().email == command.settings().user_email() };
+
+    let to_sign: IndexSet<Commit> = workspace_command
+        .parse_union_revsets(ui, &args.revisions)?
+        .evaluate_to_commits()?
+        .try_collect()?;
+
+    workspace_command.check_rewritable(to_sign.iter().ids())?;
+
+    let mut tx = workspace_command.start_transaction();
+
+    let mut signed_commits = vec![];
+    let mut foreign_commits = vec![];
+    let mut num_reparented = 0;
+
+    tx.repo_mut().transform_descendants(
+        to_sign.iter().ids().cloned().collect_vec(),
+        |rewriter| {
+            let old_commit = rewriter.old_commit().clone();
+            let commit_builder = rewriter.reparent();
+
+            if to_sign.contains(&old_commit) {
+                let new_commit = commit_builder
+                    .set_sign_key(args.key.clone())
+                    .set_sign_behavior(SignBehavior::Own)
+                    .write()?;
+
+                if !is_authored_by_me(&old_commit) {
+                    foreign_commits.push(new_commit.clone());
+                }
+
+                signed_commits.push(new_commit);
+            } else {
+                commit_builder.write()?;
+                num_reparented += 1;
+            }
+
+            Ok(())
+        },
+    )?;
+
+    if let Some(mut formatter) = ui.status_formatter() {
+        if !foreign_commits.is_empty() {
+            let template = tx.commit_summary_template();
+            writeln!(
+                ui.warning_default(),
+                "Signed {} commit(s) not authored by you",
+                foreign_commits.len()
+            )?;
+            for commit in &foreign_commits {
+                write!(formatter, "  ")?;
+                template.format(commit, formatter.as_mut())?;
+                writeln!(formatter)?;
+            }
+        }
+
+        if !signed_commits.is_empty() {
+            let template = tx.commit_summary_template();
+            writeln!(formatter, "Signed {} commit(s):", signed_commits.len())?;
+            for commit in &signed_commits {
+                write!(formatter, "  ")?;
+                template.format(commit, formatter.as_mut())?;
+                writeln!(formatter)?;
+            }
+        }
```

Comment:
It seems odd that "foreign" commits are printed twice. Maybe we can reorganize the status/warning messages something like this?
```
Signed N commits:
  ...
Warning: M of these commits are not authored by you.
```